clippy_lints/src/methods/iter_skip_next.rs

```diff
@@ -11,6 +14,23 @@ use super::ITER_SKIP_NEXT;
 pub(super) fn check(cx: &LateContext<'_>, expr: &hir::Expr<'_>, recv: &hir::Expr<'_>, arg: &hir::Expr<'_>) {
     // lint if caller of skip is an Iterator
     if is_trait_method(cx, expr, sym::Iterator) {
+        if_chain! {
+            if let Some(id) = path_to_local(recv);
+            if let Node::Binding(pat) = cx.tcx.hir().get(id);
+            if let PatKind::Binding(ann, _, _, _)  = pat.kind;
+            if ann != BindingAnnotation::Mutable;
+            then {
+                span_lint_and_note(
+                    cx,
+                    ITER_SKIP_NEXT,
+                    expr.span.trim_start(recv.span).unwrap(),
+                    "called `skip(..).next()` on an iterator",
+                    Some(pat.span),
+                    &format!("can use `nth(..)` if change {} to mut", snippet_opt(cx, pat.span).unwrap()),
```

Comment:
You could also add the `arg.span` in the help message, like it is done below. But I would check if `arg.span` is a multiline span first (if it is, I would just print the `..`). You can do this with `x.sess().source_map().is_multiline(arg.span);`

I would also split this up with `span_lint_and_then` and use `span_help` for the note to make the binding mutable and `span_suggestion` for the `.nth` part.

Reply:
> You could also add the `arg.span` in the help message, like it is done below. But I would check if `arg.span` is a multiline span first (if it is, I would just print the `..`). You can do this with `x.sess().source_map().is_multiline(arg.span);`
> 
> I would also split this up with `span_lint_and_then` and use `span_help` for the note to make the binding mutable and `span_suggestion` for the `.nth` part.

Done. Thanks.